source/extensions/common/aws/credential_providers/iam_roles_anywhere_x509_credentials_provider.cc

```diff
@@ -0,0 +1,265 @@
+  auto expired = (now - last_updated_ > REFRESH_INTERVAL);
+
+  if (expiration_time_.has_value()) {
+    return expired || (expiration_time_.value() - now < REFRESH_GRACE_PERIOD);
+  } else {
+    return expired;
+  }
+}
+
+absl::Status IAMRolesAnywhereX509CredentialsProvider::pemToAlgorithmSerialExpiration(
+    std::string pem, X509Credentials::PublicKeySignatureAlgorithm& algorithm, std::string& serial,
+    SystemTime& time) {
+  ASN1_INTEGER* ser = nullptr;
+  BIGNUM* bnser = nullptr;
+  char* bndec = nullptr;
+  absl::Status status = absl::OkStatus();
+
+  auto pemstr = pem.c_str();
+  auto pemsize = pem.size();
+  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(pemstr, pemsize));
+
+  // Not checking return code - we've already validated this certificate in previous call during der
+  // conversion
+  bssl::UniquePtr<X509> cert(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
+  X509_ALGOR* alg;
+  X509_PUBKEY_get0_param(nullptr, nullptr, nullptr, &alg, X509_get_X509_PUBKEY(cert.get()));
+  int nid = OBJ_obj2nid(alg->algorithm);
+
+  if (nid != NID_undef && nid == NID_rsaEncryption) {
+    algorithm = X509Credentials::PublicKeySignatureAlgorithm::RSA;
+  } else if (nid != NID_undef && nid == NID_X9_62_id_ecPublicKey) {
+    algorithm = X509Credentials::PublicKeySignatureAlgorithm::ECDSA;
+  } else {
+    status = absl::InvalidArgumentError("Invalid certificate public key signature algorithm");
+  }
+  ser = X509_get_serialNumber(cert.get());
+  bnser = ASN1_INTEGER_to_BN(ser, nullptr);
+  bndec = BN_bn2dec(bnser);
+  serial.append(bndec);
+
+  int days, seconds;
+
+  int rc = ASN1_TIME_diff(&days, &seconds,
+                          &Envoy::Extensions::TransportSockets::Tls::Utility::epochASN1Time(),
+                          X509_get0_notAfter(cert.get()));
+  ASSERT(rc == 1);
+  // Casting to <time_t (64bit)> to prevent multiplication overflow when certificate not-after date
+  // beyond 2038-01-19T03:14:08Z.
+  time = std::chrono::system_clock::from_time_t(static_cast<time_t>(days) * 24 * 60 * 60 + seconds);
+
+  OPENSSL_free(bndec);
+  BN_free(bnser);
+
+  return status;
+}
+
+/*TODO: @nbaws split this method and move common functionality into source/common/tls/utility.h */
+
+absl::Status IAMRolesAnywhereX509CredentialsProvider::pemToDerB64(std::string pem,
+                                                                  std::string& output, bool chain) {
+  absl::Status status = absl::OkStatus();
+
+  auto pemstr = pem.c_str();
+  auto pemsize = pem.size();
+  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(pemstr, pemsize));
+
+  auto max_certs = 1;
+  if (chain) {
+    // Maximum trust chain depth is 5 per
+    // https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication.html
+    max_certs = 5;
+  }
+  for (int i = 0; i < max_certs; i++) {
+    unsigned char* cert_in_der = nullptr;
+    bssl::UniquePtr<X509> cert(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
+
+    if (cert == nullptr) {
+      break;
```

Comment:
I wonder if this is intentional?
Specifically, what will happen if one cert is ok, and the rest are invalid (read fails)?